pkg/util/field/field_array.go

```diff
@@ -165,3 +165,76 @@ func (p *FrElementArray) String() string {
 
 	return sb.String()
 }
+
+// BatchInvert efficiently inverts the list of elements s, in place.
+func BatchInvert[T Element[T]](s []T) {
+	if len(s) == 0 {
+		return
+	}
+
+	var zero T
+	one := zero.AddUint32(1)
+
+	isZero := newBoolSlice(len(s))
+
+	m := make([]T, len(s)) // m[i] = s[i] * s[i+1] * ...
+
+	isZero.set(len(s)-1, s[len(s)-1].IsZero())
+
+	if isZero.get(len(s) - 1) {
+		s[len(s)-1] = one
+	}
+
+	m[len(s)-1] = s[len(s)-1]
+
+	for i := len(s) - 2; i >= 0; i-- {
+		isZero.set(i, s[i].IsZero())
+
+		if isZero.get(i) {
+			s[i] = one
+		}
+
+		m[i] = m[i+1].Mul(s[i])
+	}
+
+	inv := m[0].Inverse() // inv = s[0]⁻¹ * s[1]⁻¹ * ...
+
+	for i := range len(s) - 1 {
+		// inv = s[i]⁻¹ * s[i+1]⁻¹ * ...
+		s[i], inv = inv.Mul(m[i+1]), inv.Mul(s[i])
+		// inv = s[i+1]⁻¹ * s[i+2]⁻¹ * ...
+		if isZero.get(i) {
+			s[i] = zero
+		}
+	}
+
+	s[len(s)-1] = inv
+	if isZero.get(len(s) - 1) {
+		s[len(s)-1] = zero
+	}
+}
+
+// boolSlice compactly represents a boolean slice, low bytes and bits first.
+// It does not strictly enforce its size.
+type boolSlice struct {
```

Comment:
I think you can use an existing type `util/collection/bit/bit_set.go` for this.  You might like to add a `Set(bool)` method to it, that's fine.

Reply:
Great! I'll use that instead.